options/rtdl/generic/linker.cpp

```diff
@@ -1427,217 +1436,131 @@ void Loader::_scheduleInit(SharedObject *object) {
-				mlibc::infoLogger() << "rtdl: Warning: TLS_TPREL with no symbol"
-						" in object " << object->name << frg::endlog;
-			if(object->tlsModel != TlsModel::initial)
-				mlibc::panicLogger() << "rtdl: In object " << object->name
+						" in object " << rel.object()->name << frg::endlog;
+			if(rel.object()->tlsModel != TlsModel::initial)
+				mlibc::panicLogger() << "rtdl: In object " << rel.object()->name
 						<< ": Static TLS relocation to dynamically loaded object "
-						<< object->name << frg::endlog;
-			*((uint64_t *)rel_addr) = object->tlsOffset;
-		}
-	} break;
-#elif defined(__riscv)
-	case R_RISCV_64: {
-		__ensure(symbol_index);
-		__ensure(p);
-		uint64_t symbol_addr = p ? p->virtualAddress() : 0;
-		*((uint64_t *)rel_addr) = symbol_addr + reloc->r_addend;
-	} break;
-	case R_RISCV_JUMP_SLOT: {
-		__ensure(symbol_index);
-		__ensure(p);
-		__ensure(!reloc->r_addend);
-		uint64_t symbol_addr = p ? p->virtualAddress() : 0;
-		*((uint64_t *)rel_addr) = symbol_addr;
-	} break;
-	case R_RISCV_RELATIVE: {
-		__ensure(!symbol_index);
-		*((uint64_t *)rel_addr) = object->baseAddress + reloc->r_addend;
-	} break;
-	case R_RISCV_TLS_DTPMOD64: {
-		__ensure(!reloc->r_addend);
-		if(symbol_index) {
-			__ensure(p);
-			*((uint64_t *)rel_addr) = (uint64_t)p->object();
-		}else{
-			if(stillSlightlyVerbose)
-				mlibc::infoLogger() << "rtdl: Warning: TLS_DTPMOD64 with no symbol"
-						" in object " << object->name << frg::endlog;
-			*((uint64_t *)rel_addr) = (uint64_t)object;
+						<< rel.object()->name << frg::endlog;
+			tls_offset = rel.object()->tlsOffset;
 		}
-	} break;
-	case R_RISCV_TLS_DTPREL64: {
-		__ensure(symbol_index);
-		__ensure(p);
-		*((uint64_t *)rel_addr) = p->symbol()->st_value + reloc->r_addend - TLS_DTV_OFFSET;
-	} break;
-	case R_RISCV_TLS_TPREL64: {
-		// Note: this assumes that tcb_ptr + sizeof(Tcb) == tp.
-		if(symbol_index) {
-			__ensure(p);
-			__ensure(!reloc->r_addend);
-			if(p->object()->tlsModel != TlsModel::initial)
-				mlibc::panicLogger() << "rtdl: In object " << object->name
-						<< ": Static TLS relocation to dynamically loaded object "
-						<< p->object()->name << frg::endlog;
-			*((uint64_t *)rel_addr) = p->object()->tlsOffset - sizeof(Tcb) + p->symbol()->st_value;
-		}else{
-			__ensure(!reloc->r_addend);
-			if(stillSlightlyVerbose)
-				mlibc::infoLogger() << "rtdl: Warning: TLS_TPREL64 with no symbol"
-						" in object " << object->name << frg::endlog;
-			if(object->tlsModel != TlsModel::initial)
-				mlibc::panicLogger() << "rtdl: In object " << object->name
-						<< ": Static TLS relocation to dynamically loaded object "
-						<< object->name << frg::endlog;
-			*((uint64_t *)rel_addr) = object->tlsOffset - sizeof(Tcb);
+
+		if constexpr (tlsAboveTp) {
+			off += tls_offset - sizeof(Tcb);
+		} else {
+			off += tls_offset;
 		}
+
+		rel.relocate(off);
 	} break;
-	case R_RISCV_NONE:
-		// Work around https://sourceware.org/bugzilla/show_bug.cgi?id=24673.
```

Comment:
What about this workaround?

Reply:
This is now handled by `R_NONE`.

Reply:
Oh must've missed it, my bad. Perhaps keep the comment saying that it works around a binutils bug on RV?

Reply:
The issue is still linked where it's relevant (TEXTREL handling), so I removed this here as ignoring R_NONE-type relocations is correct behavior anyways and not really relevant to handling the binutils bug outside of that.

Reply:
Is it actually correct to ignore none-type relocs? Feels like we might want to at least warn on those

Reply:
Every libc I checked just ignores them and the ABI specs all just say "perform no relocation", no more, no less.

Reply:
It probably says that for forward compatability reasons, but IMHO it could be good to warn because it could indicate a problem with the toolchain or linker